File: server/app/models/property.py
from app import db

class Property(db.Model):
    __tablename__ = 'properties'

    id = db.Column(db.Integer, primary_key=True)
    host_id = db.Column(db.Integer, db.ForeignKey('users.id'), nullable=False)
    title = db.Column(db.String(200), nullable=False)
    description = db.Column(db.Text, nullable=False)
    price_per_night = db.Column(db.Numeric(10, 2), nullable=False)
    location = db.Column(db.String(255), nullable=False)
    photos = db.Column(db.JSON, nullable=False)
    amenities = db.Column(db.JSON)
    is_active = db.Column(db.Boolean, default=True)
    created_at = db.Column(db.DateTime, default=db.func.current_timestamp())

    bookings = db.relationship('Booking', backref='property', lazy='dynamic')

    def to_dict(self):
        return {
            'id': self.id,
            'host_id': self.host_id,
            'host_name': self.host.name if self.host else None,
            'title': self.title,
            'description': self.description,
            'price_per_night': float(self.price_per_night),
            'location': self.location,
            'photos': self.photos,
            'amenities': self.amenities or [],
            'is_active': self.is_active,
            'created_at': self.created_at.isoformat() if self.created_at else None,
        }

File: server/src/models/Property.ts
import mongoose, { Schema, Document } from 'mongoose';

export interface IProperty extends Document {
  host: mongoose.Types.ObjectId;
  title: string;
  description: string;
  pricePerNight: number;
  location: string;
  photos: string[];
  amenities?: string[];
  unavailableDates?: Date[];
  isActive: boolean;
  createdAt: Date;
}

const propertySchema = new Schema<IProperty>({
  host: { type: Schema.Types.ObjectId, ref: 'User', required: true },
  title: { type: String, required: true, trim: true },
  description: { type: String, required: true },
  pricePerNight: { type: Number, required: true, min: 0 },
  location: { type: String, required: true, trim: true },
  photos: {
    type: [String],
    validate: { validator: (v: string[]) => v.length > 0, message: 'At least one photo required' },
  },
  amenities: [{ type: String }],
  unavailableDates: [{ type: Date }],
  isActive: { type: Boolean, default: true },
  createdAt: { type: Date, default: Date.now },
});

export default mongoose.model<IProperty>('Property', propertySchema);
